src/lib.rs

```diff
@@ -98,11 +136,88 @@ pub struct BaseState {
     request_focus: bool,
 }
 
+/// The trait implemented by all widgets.
+///
+/// All appearance and behavior for a widget is encapsulated in an
+/// object that implements this trait.
+///
+/// The trait is parametrized by a type (`T`) for associated data.
+/// All trait methods are provided with access to this data, and
+/// in the case of `event` the reference is mutable, so that events
+/// can directly update the data.
+///
+/// Whenever the application data changes, the framework traverses
+/// the widget hierarchy with an [`update`] method. The framework
+/// needs to know whether the data has actually changed or not, which
+/// is why `T` has a [`Data`] bound.
+///
+/// All the trait methods are provided with a corresponding context.
+/// The widget can request things and cause actions by calling methods
+/// on that context.
+///
+/// In addition, all trait methods are provided with an environment
+/// ([`Env`](struct.Env.html)).
+///
+/// Container widgets will generally not call `Widget` methods directly
+/// on their child widgets, but rather will own their widget wrapped in
+/// a [`WidgetPod`], and call the corresponding method on that. The
+/// `WidgetPod` contains state and logic for these traversals. On the
+/// other hand, particularly light-weight containers might contain their
+/// child `Widget` directly (when no layout or event flow logic is
+/// needed), and in those cases will call these methods.
+///
+/// As a general pattern, container widgets will call the corresponding
+/// `WidgetPod` method on all their children. The `WidgetPod` applies
+/// logic to determine whether to recurse, as needed.
+///
+/// [`event`]: #tymethod.event
+/// [`update`]: #tymethod.update
+/// [`Data`]: trait.Data.html
+/// [`WidgetPod`]: struct.WidgetPod.html
 pub trait Widget<T> {
+    /// Paint the widget appearance.
+    ///
+    /// The widget calls methods on the `render_target` field of the
```

Comment:
did this change from `render_ctx`?

As discussed previously I think `PaintCtx` should deref to `RenderCtx`, or otherwise re-expose the required functionality, but let's get this merged without any more changes.

Reply:
Yeah, I think it should deref. Let's get that in shortly, but as you remark it feels slightly out of scope for just a merge (especially as there's a fair amount of cleanup in the uses in paint methods).